File: src/main/java/com/colombia/eps/savesession/repository/DynamoRepository.java
package com.colombia.eps.savesession.repository;

import com.colombia.eps.savesession.helper.exception.CountSessionByIdException;
import com.colombia.eps.savesession.helper.exception.DeleteOldestSessionException;
import com.colombia.eps.savesession.helper.exception.GetOldestSessionException;
import com.colombia.eps.savesession.helper.exception.SaveSessionInTableException;
import com.colombia.eps.savesession.helper.exception.SaveSessionProcessException;
import com.colombia.eps.savesession.model.Session;
import jakarta.annotation.PostConstruct;
import jakarta.enterprise.context.ApplicationScoped;
import lombok.RequiredArgsConstructor;
import org.jboss.logging.Logger;
import software.amazon.awssdk.enhanced.dynamodb.DynamoDbEnhancedClient;
import software.amazon.awssdk.enhanced.dynamodb.DynamoDbTable;
import software.amazon.awssdk.enhanced.dynamodb.Key;
import software.amazon.awssdk.enhanced.dynamodb.TableSchema;
import software.amazon.awssdk.enhanced.dynamodb.model.QueryConditional;
import software.amazon.awssdk.enhanced.dynamodb.model.QueryEnhancedRequest;
import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
import software.amazon.awssdk.services.dynamodb.model.*;

import java.util.Map;
import java.util.Optional;

@RequiredArgsConstructor
@ApplicationScoped
public class DynamoRepository {

    private final DynamoDbClient dynamoDbClient;
    private DynamoDbTable<Session> sessionTable;
    private static final Logger LOG = Logger.getLogger(DynamoRepository.class);

    @PostConstruct
    public void init() {
        String tableName = System.getenv("DYNAMO_TABLE_NAME");
        LOG.debug(String.format("Inicializando DynamoRepository con tabla: %s", tableName));

        // Inicializa el cliente mejorado
        DynamoDbEnhancedClient enhancedClient = DynamoDbEnhancedClient.builder()
                .dynamoDbClient(dynamoDbClient)
                .build();

        // Configura la tabla
        this.sessionTable = enhancedClient.table(tableName, TableSchema.fromBean(Session.class));
    }

    public void saveSession(String sessionId, String ip, String city, String timestamp) {
        try {
            int sessionCount = countSessionsById(sessionId);
            LOG.debug(String.format("Cantidad de sesiones encontradas: %s", sessionCount));
            saveInTable(sessionId, ip, city, timestamp);
            if (sessionCount > 5) {
                LOG.debug("Límite alcanzado, eliminando sesión más antigua...");
                deleteOldestSession(sessionId);
                LOG.debug("Sesión más antigua eliminada");
            }
        } catch (Exception exception) {
            LOG.error(String.format("Error en el proceso de guardar sesión: %s", exception.getMessage()));
            throw new SaveSessionProcessException();
        }
    }

    private void saveInTable(String sessionId, String ip, String city, String timestamp) {
        try {
            LOG.debug("Creando item para DynamoDB");
            Session session = Session.builder()
                    .email(sessionId)
                    .connectionTime(timestamp)
                    .ip(ip)
                    .city(city)
                    .build();

            LOG.debug("Enviando item a DynamoDB");
            sessionTable.putItem(session);  // Usamos putItem con el cliente mejorado
            LOG.debug("Item guardado exitosamente en DynamoDB");
        } catch (Exception exception) {
            LOG.error(String.format("Error al guardar la sesión en la tabla: %s", exception.getMessage()));
            throw new SaveSessionInTableException();
        }
    }

    private int countSessionsById(String sessionId) {
        try {
            LOG.debug(String.format("Contando registros con sessionId: %s", sessionId));
            
            String tableName = System.getenv("DYNAMO_TABLE_NAME");
            
            QueryRequest queryRequest = QueryRequest.builder()
                    .tableName(tableName)
                    .keyConditionExpression("id = :pk")
                    .expressionAttributeValues(Map.of(":pk", AttributeValue.builder().s(sessionId).build()))
                    .select(Select.COUNT)
                    .build();

            QueryResponse response = dynamoDbClient.query(queryRequest);
            int count = response.count();
            LOG.debug(String.format("Registros encontrados con sessionId %s: %s", sessionId, count));
            return count;
        } catch (Exception e) {
            LOG.error(String.format("Error contando registros por sessionId: %s", e.getMessage()));
            throw new CountSessionByIdException();
        }
    }

    private void deleteOldestSession(String sessionId) {
        try {
            LOG.debug(String.format("Eliminando sesión más antigua para sessionId: %s", sessionId));
            Optional<Session> oldestSession = getSession(sessionId);
            if (oldestSession.isPresent()) {
                String session = oldestSession.get().toString();
                LOG.debug(String.format("Sesión más antigua encontrada: %s", session));
                sessionTable.deleteItem(oldestSession.get());  // Usamos deleteItem con el cliente mejorado
                LOG.debug("Proceso de eliminación de la sesión más antigua exitoso");
            }
        } catch (Exception e) {
            LOG.error(String.format("Error eliminando sesión más antigua: %s", e.getMessage()));
            throw new DeleteOldestSessionException();
        }
    }

    private Optional<Session> getSession(String sessionId) {
        try {
            LOG.debug(String.format("Obteniendo sesión más antigua para sessionId: %s", sessionId));
            
            QueryConditional queryConditional = QueryConditional.keyEqualTo(Key.builder()
                    .partitionValue(sessionId)
                    .build());

            QueryEnhancedRequest queryRequest = QueryEnhancedRequest.builder()
                    .queryConditional(queryConditional)
                    .scanIndexForward(true)  // true = ascending (oldest first)
                    .limit(1)
                    .build();

            return sessionTable.query(queryRequest).items().stream().findFirst();
        } catch (Exception e) {
            LOG.error(String.format("Error obteniendo sesión más antigua: %s", e.getMessage()));
            throw new GetOldestSessionException();
        }
    }
}

File: src/repository/dynamoRepository.js
/**
 * DynamoDB Repository
 */

const { DynamoDBDocumentClient, PutCommand, QueryCommand, DeleteCommand } = require('@aws-sdk/lib-dynamodb');
const Session = require('../model/session');
const CONSTANTS = require('../helpers/constants');
const { handleError } = require('../helpers/auxiliaryMethods');

class DynamoRepository {
    constructor(dynamoClient) {
        this.docClient = DynamoDBDocumentClient.from(dynamoClient);
        this.tableName = process.env.DYNAMO_TABLE_NAME;

        if (!this.tableName) {
            throw new Error('DYNAMO_TABLE_NAME environment variable is required');
        }

    }

    /**
     * Save a session directly in the table
     * @param {string} sessionId - Session ID
     * @param {string} ip - IP address
     * @param {string} city - City name
     * @param {string} country - Country
     * @param {string} timezone - Timezone name
     * @param {string} coordinates - Coordinates
     * @param {string} timestamp - Timestamp of the session
     */
    async saveSession(sessionId, ip, city, timezone, country, coordinates) {
        try {

            const session = new Session(sessionId, ip, city, timezone, country, coordinates);
            console.log(`Item creado: ${session.toString()}`);
            if (!session.isValid()) {
                throw new Error('Session data is invalid');
            }

            console.log('Enviando item a DynamoDB');

            const params = {
                TableName: this.tableName,
                Item: session.toDynamoItem()
            };

            await this.docClient.send(new PutCommand(params));
            console.log('Item guardado exitosamente en DynamoDB');

        } catch (error) {
            throw handleError('saveSession', error);
        }
    }

    /**
     * Cuenta las sesiones existentes para un sessionId
     * @param {string} sessionId - ID de la sesión
     * @returns {Promise<number>} - Cantidad de sesiones
     */
    async countSessionsById(sessionId) {
        try {
            console.log(`Contando registros con sessionId: ${sessionId}`);

            const params = {
                TableName: this.tableName,
                KeyConditionExpression: 'id = :pk',
                ExpressionAttributeValues: {
                    ':pk': sessionId
                },
                Select: 'COUNT'
            };

            const result = await this.docClient.send(new QueryCommand(params));
            const count = result.Count;

            console.log(`Registros encontrados con sessionId ${sessionId}: ${count}`);
            return count;
        } catch (error) {
            throw handleError('countSessionsById', error);
        }
    }

    /**
     * Delete oldest session for a given sessionId
     * @param {string} sessionId - ID of the session to delete
     */
    async deleteOldestSession(sessionId) {
        try {
            console.log(`Eliminando sesión más antigua para sessionId: ${sessionId}`);

            const oldestSession = await this.getOldestSession(sessionId);
            if (oldestSession) {
                console.log(`Sesión más antigua encontrada: ${JSON.stringify(oldestSession)}`);

                const params = {
                    TableName: this.tableName,
                    Key: {
                        id: oldestSession.id,
                        connection_time: oldestSession.connection_time
                    }
                };

                await this.docClient.send(new DeleteCommand(params));
                console.log('Proceso de eliminación de la sesión más antigua exitoso');
            }
        } catch (error) {
            throw handleError('deleteOldestSession', error);
        }
    }

    /**
     * Ge la sesión más antigua para un sessionId
     * @param {string} sessionId - ID de la sesión
     * @returns {Promise<Object|null>} - Sesión más antigua o null
     */
    async getOldestSession(sessionId) {
        try {
            console.log(`Obteniendo sesión más antigua para sessionId: ${sessionId}`);

            const params = {
                TableName: this.tableName,
                KeyConditionExpression: 'id = :pk',
                ExpressionAttributeValues: {
                    ':pk': sessionId
                },
                ScanIndexForward: true,
                Limit: 1
            };

            const result = await this.docClient.send(new QueryCommand(params));
            return result.Items && result.Items.length > 0 ? result.Items[0] : null;
        } catch (error) {
            throw handleError('getOldestSession', error);
        }
    }
}

module.exports = DynamoRepository;